flytepropeller/pkg/resolver/k8s_resolver.go

```diff
@@ -0,0 +1,182 @@
+func NewBuilder(client *kubernetes.Clientset, schema string) resolver.Builder {
+	return &kubeBuilder{
+		k8sClient: client,
+		schema:    schema,
+	}
+}
+
+type kubeBuilder struct {
+	k8sClient *kubernetes.Clientset
+	schema    string
+}
+
+func splitServicePortNamespace(hpn string) (service, port, namespace string) {
+	service = hpn
+
+	colon := strings.LastIndexByte(service, ':')
+	if colon != -1 {
+		service, port = service[:colon], service[colon+1:]
+	}
+
+	// we want to split into the service name, namespace, and whatever else is left
+	// this will support fully qualified service names, e.g. {service-name}.<namespace>.svc.<cluster-domain-name>.
+	// Note that since we look up the endpoints by service name and namespace, we don't care about the
+	// cluster-domain-name, only that we can parse out the service name and namespace properly.
+	parts := strings.SplitN(service, ".", 3)
+	if len(parts) >= 2 {
+		service, namespace = parts[0], parts[1]
+	}
+
+	return
+}
+
+func parseResolverTarget(target resolver.Target) (targetInfo, error) {
+	var service, port, namespace string
+	if target.URL.Host == "" {
+		// kubernetes:///service.namespace:port
+		service, port, namespace = splitServicePortNamespace(target.Endpoint())
+	} else if target.URL.Port() == "" && target.Endpoint() != "" {
+		// kubernetes://namespace/service:port
+		service, port, _ = splitServicePortNamespace(target.Endpoint())
+		namespace = target.URL.Hostname()
+	} else {
+		// kubernetes://service.namespace:port
+		service, port, namespace = splitServicePortNamespace(target.URL.Host)
+	}
+
+	if service == "" {
+		return targetInfo{}, fmt.Errorf("target %s must specify a service", &target.URL)
+	}
+
+	return targetInfo{
+		serviceName:      service,
+		serviceNamespace: namespace,
+		port:             port,
+	}, nil
+}
+
+// Build creates a new resolver for the given target, e.g. k8s:///flyteagent:flyte:8000.
+func (b *kubeBuilder) Build(target resolver.Target, cc resolver.ClientConn, opts resolver.BuildOptions) (resolver.Resolver, error) {
+	ti, err := parseResolverTarget(target)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	r := &kResolver{
+		target:    ti,
+		ctx:       ctx,
+		cancel:    cancel,
+		cc:        cc,
+		k8sClient: b.k8sClient,
+	}
+	go wait.Until(r.run, 0, ctx.Done())
+	return r, nil
+}
+
+// Scheme returns the scheme supported by this resolver.
+// Scheme is defined at https://github.com/grpc/grpc/blob/master/doc/naming.md.
+func (b *kubeBuilder) Scheme() string {
+	return b.schema
+}
+
+type kResolver struct {
+	target    targetInfo
+	ctx       context.Context
+	cancel    context.CancelFunc
+	cc        resolver.ClientConn
+	k8sClient *kubernetes.Clientset
+	// wg is used to enforce Close() to return after the watcher() goroutine has finished.
+	wg sync.WaitGroup
+}
+
+// ResolveNow is a no-op at this point.
+func (k *kResolver) ResolveNow(resolver.ResolveNowOptions) {
```

Comment:
We want to implement this, right?

Reply:
is it necessary? watcher should be able to update the endpoint immediately. xds resolver doesn't implement this, either.
https://chromium.googlesource.com/external/github.com/grpc/grpc-go/+/HEAD/xds/internal/resolver/xds_resolver.go#308

Reply:
It's not necessary. Just to confirm. The watching mechanism makes it unnecessary and useless anyway.